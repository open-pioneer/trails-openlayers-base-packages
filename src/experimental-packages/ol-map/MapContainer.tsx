// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { Box } from "@open-pioneer/chakra-integration";
import classNames from "classnames";
import type OlMap from "ol/Map";
import { Extent } from "ol/extent";
import { useService } from "open-pioneer:react-hooks";
import { ReactNode, useEffect, useMemo, useRef } from "react";
import { useAsync } from "react-use";
import { MapContextProvider, MapContextType } from "./MapContext";

export interface OlComponentProps {
    mapId: string;
}

/**
 * Map padding, all values are pixels.
 *
 * See https://openlayers.org/en/latest/apidoc/module-ol_View-View.html#padding
 */
export interface MapPadding {
    left?: number;
    right?: number;
    top?: number;
    bottom?: number;
}

export interface MapComponentProps extends OlComponentProps {
    /**
     * Sets the map's padding directly.
     *
     * See: https://openlayers.org/en/latest/apidoc/module-ol_View-View.html#padding)
     */
    viewPadding?: MapPadding | undefined;

    /**
     * Behavior performed by the map when the view padding changes.
     *
     * - `none`: Do nothing.
     * - `preserve-center`: Ensures that the center point remains the same by animating the view.
     * - `preserve-extent`: Ensures that the extent remains the same by zooming.
     *
     * @default "preserve-center"
     */
    viewPaddingChangeBehavior?: "none" | "preserve-center" | "preserve-extent";

    /**
     * Additional class name(s).
     */
    className?: string;

    children?: ReactNode;
}

export function MapContainer(props: MapComponentProps) {
    const { mapId, className, ...rest } = props;

    const mapElement = useRef<HTMLDivElement>(null);
    const mapRegistry = useService("ol-map.MapRegistry");
    const mapState = useAsync(async () => await mapRegistry.getMap(mapId), [mapId]);
    const map = mapState.value; // undefined -> not ready yet

    useEffect(() => {
        if (map && mapElement.current) {
            // Register div as render target
            const resource = mapRegistry.setContainer(mapId, mapElement.current);
            return () => resource.destroy();
        }
    }, [map, mapRegistry, mapId]);

    return (
        <Box className={classNames("map-container", className)} ref={mapElement} h="100%" w="100%">
            {map && <MapContainerReady map={map} {...rest} />}
        </Box>
    );
}

/**
 * This inner component is rendered when the map has been loaded.
 *
 * It provides the map instance and additional properties down the component tree.
 */
function MapContainerReady(
    props: { map: OlMap } & Omit<MapComponentProps, "mapId" | "className">
): JSX.Element {
    const {
        map,
        viewPadding: viewPaddingProp,
        viewPaddingChangeBehavior = "preserve-center",
        children
    } = props;

    const viewPadding = useMemo<Required<MapPadding>>(() => {
        return {
            left: viewPaddingProp?.left ?? 0,
            right: viewPaddingProp?.right ?? 0,
            top: viewPaddingProp?.top ?? 0,
            bottom: viewPaddingProp?.bottom ?? 0
        };
    }, [viewPaddingProp]);

    // Apply view padding
    useEffect(() => {
        const mapView = map?.getView();
        if (!map || !mapView) {
            return;
        }

        const oldCenter = mapView.getCenter();
        const oldPadding = fromOlPadding(mapView.padding);
        const oldExtent = extentIncludingPadding(map, oldPadding);

        mapView.padding = toOlPadding(viewPadding);
        switch (viewPaddingChangeBehavior) {
            case "preserve-center":
                mapView.animate({ center: oldCenter, duration: 300 });
                break;
            case "preserve-extent": {
                if (oldExtent) {
                    mapView.animate({
                        center: oldCenter,
                        resolution: mapView.getResolutionForExtent(oldExtent),
                        duration: 300
                    });
                }
                break;
            }
            case "none":
        }
    }, [viewPadding, map, viewPaddingChangeBehavior]);

    const mapContext = useMemo((): MapContextType => {
        return {
            map,
            padding: viewPadding
        };
    }, [map, viewPadding]);
    return <MapContextProvider value={mapContext}>{children}</MapContextProvider>;
}

/**
 * Returns the extent visible in the non-padded region of the map.
 */
function extentIncludingPadding(map: OlMap, padding: Required<MapPadding>): Extent | undefined {
    const size = map.getSize();
    if (!size || size.length < 2) {
        return undefined;
    }

    const [width, height] = size as [number, number];
    const bottomLeft = map.getCoordinateFromPixel([padding.left, padding.bottom]);
    const topRight = map.getCoordinateFromPixel([
        Math.max(0, width - padding.right),
        Math.max(0, height - padding.top)
    ]);
    if (!bottomLeft || !topRight) {
        return undefined;
    }

    const [xmin, ymin] = bottomLeft;
    const [xmax, ymax] = topRight;
    return [xmin, ymin, xmax, ymax] as Extent;
}

function fromOlPadding(padding: number[] | undefined): Required<MapPadding> {
    // top, right, bottom, left
    return {
        top: padding?.[0] ?? 0,
        right: padding?.[1] ?? 0,
        bottom: padding?.[2] ?? 0,
        left: padding?.[3] ?? 0
    };
}

function toOlPadding(padding: Required<MapPadding>): number[] {
    // top, right, bottom, left
    const { top, right, bottom, left } = padding;
    return [top, right, bottom, left];
}
