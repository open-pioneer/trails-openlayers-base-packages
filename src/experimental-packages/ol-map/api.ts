// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import type { Resource } from "@open-pioneer/core";
import type OlMap from "ol/Map";
import type { MapOptions } from "ol/Map";

/**
 * Provides an open layers map configuration with a given map id.
 */
export interface OlMapConfigurationProvider {
    /**
     * Identifier of the map.
     */
    readonly mapId: string;

    /**
     * Returns the map options that will be applied on the corresponding map.
     */
    getMapOptions(): Promise<MapOptions>;
}

export interface OlMapRegistry {
    /**
     * Returns the map with the given id, initializing it if necessary.
     */
    getMap(mapId: string): Promise<OlMap>;

    /**
     * Registers the container for the given map.
     * This is used internally by the `MapContainer` component.
     *
     * There can only be (at most) one container per map.
     *
     * Call `.destroy()` on the returned resource to unregister the `target`.
     */
    setContainer(mapId: string, target: HTMLDivElement): Resource;
}

import "@open-pioneer/runtime";
declare module "@open-pioneer/runtime" {
    interface ServiceRegistry {}
}
