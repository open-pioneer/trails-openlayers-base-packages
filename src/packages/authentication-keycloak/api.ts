// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import type { KeycloakConfig, KeycloakInitOptions } from "keycloak-js";

/**
 * The central configuration properties of the plugin.
 *
 * An object of this type should be used as configuration for this package.
 */
export interface KeycloakProperties {
    /**
     * These properties are required by the Keycloak JavaScript adapter.
     */
    keycloakOptions: KeycloakOptions;
}

export interface KeycloakOptions {
    /**
     * The configuration details for connecting to Keycloak.
     * 'url': The URL of your Keycloak server.
     * 'realm': The realm within Keycloak.
     * 'clientId': The ID of the client application registered in Keycloak.
     */
    keycloakConfig: KeycloakConfig;

    /**
     * Control the automatic refreshing of authentication tokens.
     * 'autoRefresh': Whether token refreshing should happen automatically.
     * 'interval': The interval (in milliseconds) at which token refreshing should occur.
     * 'timeLeft': The remaining time (in milliseconds) before token expiration.
     */
    refreshOptions: RefreshOptions;

    /**
     * Define how Keycloak initializes.
     * This properties can be used:
     * 'onLoad': Specifies when Keycloak should initialize.
     * 'pkceMethod': The method used for PKCE for enhanced security.
     * 'scope': The scope of the authentication.
     *
     */
    keycloakInitOptions: Partial<KeycloakInitOptions>;
}

/**
 * Control the automatic refreshing of authentication tokens.
 */
export interface RefreshOptions {
    /**
     * Whether token refreshing should happen automatically.
     */
    autoRefresh: boolean;
    /**
     * The interval (in milliseconds) at which token refreshing should occur.
     */
    interval: number;
    /**
     * The remaining time (in milliseconds) before token expiration.
     */
    timeLeft: number;
}
