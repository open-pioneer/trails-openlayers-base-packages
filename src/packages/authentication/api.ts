// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import type { EventSource } from "@open-pioneer/core";
import type { DeclaredService } from "@open-pioneer/runtime";
import type { ComponentType } from "react";

/**
 * Events emitted by the {@link AuthService}.
 */
export interface AuthEvents {
    /** Emitted when there were any changes to the service's state. */
    changed: void;
}

/**
 * Information about the authenticated user's session.
 */
export interface SessionInfo {
    /** Technical user id of the authenticated user, for example an email address. */
    userId: string;

    /**
     * Display name of the authenticated user.
     * Use the {@link userId} as a default value if this value is not available.
     */
    userName?: string | undefined;

    /** Set to a date if the session expires at some point. Optional. */
    expiresAt?: Date | undefined;

    /** Arbitrary attributes from the authentication plugin. */
    attributes?: Record<string, unknown> | undefined;
}

/**
 * Models the current authentication state.
 *
 * NOTE: Future versions of this package may define additional states.
 * Your code should contain sensible fallback or error logic.
 */
export type AuthState = AuthStatePending | AuthStateNotAuthenticated | AuthStateAuthenticated;

/**
 * This state is active when the authentication service
 * is still checking whether the current user is authenticated or not.
 */
export interface AuthStatePending {
    kind: "pending";
}

/**
 * The user not authenticated.
 */
export interface AuthStateNotAuthenticated {
    kind: "not-authenticated";
}

/**
 * The user is authenticated and its session attributes
 * can be retrieved.
 */
export interface AuthStateAuthenticated {
    kind: "authenticated";
    sessionInfo: SessionInfo;
}

/**
 * Defines the behavior of the authentication service when attempting to
 * authenticate a user.
 */
export type LoginBehavior = LoginFallback | LoginEffect;

/**
 * A fallback react component to present to the user.
 * For example, this can be a login form or a message.
 */
export interface LoginFallback {
    kind: "fallback";
    Fallback: ComponentType;
}

/**
 * An effect to perform when the user shall be authenticated.
 * `login()` may, for example, perform a redirect to an authentication provider.
 */
export interface LoginEffect {
    kind: "effect";
    login(): void;
}

/**
 * Manages the current user's authentication state.
 *
 * The current state (such as session info) can be retrieved and watched for changes.
 */
export interface AuthService
    extends EventSource<AuthEvents>,
        DeclaredService<"authentication.AuthService"> {
    /**
     * Returns the current authentication state.
     *
     * The state may initially be `pending` to allow for async initialization in the authentication plugin.
     * After initialization, the state is either `not-authenticated` or `authenticated`.
     * 
      This method must be called again after the {@link AuthService} has emitted the `changed` event.
     */
    getAuthState(): AuthState;

    /**
     * Returns the current user's {@link SessionInfo} or `undefined`, if the current user is not authenticated.
     *
     * The method is asynchronous to allow for async initialization in the authentication plugin.
     *
     * This method must be called again after the {@link AuthService} has emitted the `changed` event.
     */
    getSessionInfo(): Promise<SessionInfo | undefined>;

    /**
     * Returns the login behavior that should be performed if the user is not authenticated.
     *
     * The actual implementation of this component depends on the application's authentication plugin.
     */
    getLoginBehavior(): LoginBehavior;

    /**
     * Terminates the current session (if any).
     */
    logout(): void;
}

/** Events that may be emitted by an authentication plugin. */
export interface AuthPluginEvents {
    changed: void;
}

/** Optional base type for an authentication plugin: the event emitter interface is not required. */
export type AuthPluginEventBase = EventSource<AuthPluginEvents>;

/**
 * The authentication service requires an AuthPlugin to implement a concrete authentication flow.
 *
 * The plugin provides the current authentication state and the authentication fallback to the service.
 *
 * The current authentication state returned by {@link getAuthState} may change.
 * If that is the case, the plugin must also emit the `changed` event to notify the service.
 *
 * The implementation of `AuthPluginEventBase` is optional: it is only necessary if the state changes
 * during the lifetime of the plugin.
 * To implement the event, you can write `class MyPlugin extends EventEmitter<AuthPluginEvents>`.
 */
export interface AuthPlugin
    extends Partial<AuthPluginEventBase>,
        DeclaredService<"authentication.AuthPlugin"> {
    /**
     * Returns the current authentication state.
     *
     * Objects returned by this method should not be mutated.
     * Emit the `changed` event instead to communicate that there is a new state.
     */
    getAuthState(): AuthState;

    /**
     * Returns the login behavior that should be performed if the user is not authenticated.
     */
    getLoginBehavior(): LoginBehavior;

    /**
     * Explicitly triggers a logout.
     *
     * Should result in a new state (including a `changed` event) if the user
     * was authenticated.
     */
    logout(): Promise<void> | void;
}
