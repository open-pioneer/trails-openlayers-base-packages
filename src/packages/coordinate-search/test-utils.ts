// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { act, fireEvent } from "@testing-library/react";

export function showDropdown(projSelect: HTMLElement) {
    // open dropdown to include options in snapshot; react-select creates list of options in dom after opening selection
    act(() => {
        fireEvent.keyDown(projSelect, { key: "ArrowDown" });
    });
}

export function getCurrentOptions() {
    return Array.from(
        document.getElementsByClassName("coordinate-input-select__option")
    ) as HTMLElement[];
}

export function getCurrentOptionValues(options: HTMLElement[]) {
    const values: (string | null)[] = [];
    for (const opt of options) {
        values.push(opt.textContent);
    }
    return values;
}

export function getClearButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const clearButton = buttonDiv.querySelector(".coordinate-input-clear-button");
    if (!clearButton) {
        throw new Error("clear button not rendered");
    }
    return clearButton;
}

export function getCopyButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const copyButton = buttonDiv.querySelector(".coordinate-input-copy-button");
    if (!copyButton) {
        throw new Error("copy button not rendered");
    }
    return copyButton;
}
