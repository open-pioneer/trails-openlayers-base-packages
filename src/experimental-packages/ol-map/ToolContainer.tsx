// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { Box, StyleProps } from "@open-pioneer/chakra-integration";
import classNames from "classnames";
import { ReactNode } from "react";
import { createPortal } from "react-dom";
import { useMapContext } from "./MapContext";
import { MapPadding } from "./MapContainer";

export type ToolContainerPosition = "top-left" | "top-right" | "bottom-left" | "bottom-right";

// TODO: Branch umbennen, wenn MapAPI fertig und Implementierung Ã¼bernehmen
// TODO: Doku anpassen -> Tab order tool container richtige Reihenfolge konfigurieren
// TODO: Tests

export interface ToolContainerProps {
    /**
     * The position of the container above the map.
     * @default "top-right"
     */
    position?: ToolContainerPosition;
    className?: string | undefined;
    children?: ReactNode;
    /**
     * Horizontal gap in pixel applied to ToolContainer.
     *
     * Applied:
     * - left, if position `*-left`
     * - right, if position `*-right`
     *
     * @default 0
     */
    horizontalGap?: number;
    /**
     * Vertical gap in pixel applied to ToolContainer.
     *
     * Applied:
     * - top, if position `*-top`
     * - bottom, if position `*-bottom`
     *
     * @default 0 (If position `*-bottom`, default verticalGap == `30`)
     */
    verticalGap?: number;
}

export function ToolContainer(props: ToolContainerProps): JSX.Element {
    const { position = "top-right", className, children, horizontalGap, verticalGap } = props;
    const { map, padding } = useMapContext();
    const overlayContainer = map.getOverlayContainerStopEvent();

    return createPortal(
        <Box
            className={classNames("tool-container", className)}
            /* Overlay container uses pointer-events: none, this restores interactivity */
            pointerEvents="auto"
            /* Restore user-select: none set by ol-viewport parent */
            userSelect="text"
            {...computePositionStyles(position, padding, horizontalGap, verticalGap)}
        >
            {children}
        </Box>,
        overlayContainer
    );
}

export function computePositionStyles(
    position: ToolContainerPosition,
    padding: Required<MapPadding>,
    horizontalGap?: number | undefined,
    verticalGap?: number | undefined
): StyleProps {
    const props: StyleProps = {
        position: "absolute",
        transitionProperty: "left, right, top, bottom",
        transitionDuration: "200ms",
        transitionTimingFunction: "ease-out"
    };

    const defaultHorizontalGap = 0;
    const horizontal = horizontalGap ?? defaultHorizontalGap;

    const defaultVerticalGap = 0;
    const vertical = verticalGap ?? defaultVerticalGap;

    /**
     * improvement: Get height directly from `Attribution` HTMLDivElement
     */
    const attributionHeight = 20;
    const attributionSpace = 10;
    const attributionGap = verticalGap === undefined ? attributionHeight + attributionSpace : 0;

    const gap = (n: number) => `${n}px`;

    switch (position) {
        case "top-left":
            props.left = gap(padding.left + horizontal);
            props.top = gap(padding.top + vertical);
            break;
        case "top-right":
            props.right = gap(padding.right + horizontal);
            props.top = gap(padding.top + vertical);
            break;
        case "bottom-left":
            props.left = gap(padding.left + horizontal);
            props.bottom = gap(padding.bottom + vertical + attributionGap);
            break;
        case "bottom-right":
            props.right = gap(padding.right + horizontal);
            props.bottom = gap(padding.bottom + vertical + attributionGap);
            break;
    }

    /**
     * Apply max-height and max-width to ToolContainer to avoid content overflow
     */
    props.maxH = `calc((100%) - ${props.top ?? "0px"} - ${
        props.bottom ?? attributionGap + "px"
    } - ${vertical + "px"} - ${attributionSpace + "px"})`;

    props.maxW = `calc((100%) - ${props.left ?? "0px"} - ${props.right ?? "0px"} - ${
        horizontal + "px"
    })`;
    props.overflow = "hidden";

    return props;
}
