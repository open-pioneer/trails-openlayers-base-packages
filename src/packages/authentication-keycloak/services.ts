// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import Keycloak, {
    KeycloakInitOptions,
    KeycloakLoginOptions,
    KeycloakLogoutOptions
} from "keycloak-js";
import { KeycloakConfigProvider } from "./api";
/*export const keycloak = new Keycloak({
    url: "https://auth.ldproxy.net/",
    realm: "ii",
    clientId: "it-nrw"
});

export const keycloak = new Keycloak({
    url: "https://hsi-pex0-13620.service.it.nrw.de/keycloak/",
    realm: "trails",
    clientId: "trails-sample"
});
*/
let uKeycloak = new Keycloak();

export class KeycloakConfigProviderImpl implements KeycloakConfigProvider {
    setKeycloak(keycloak: Keycloak) {
        uKeycloak = keycloak;
        return keycloak;
    }

    getInitOptions(
        options: KeycloakInitOptions | undefined = {
            onLoad: "check-sso",
            pkceMethod: "S256",
            scope: "data:read"
        }
    ) {
        return options;
    }

    getLoginOptions(options: KeycloakLoginOptions | undefined = { redirectUri: undefined }) {
        return options;
    }

    getLogoutOptions(options: KeycloakLogoutOptions | undefined = { redirectUri: undefined }) {
        return options;
    }

    getRefreshOptions(
        options: { autoRefresh: boolean; interval: number; timeLeft: number } | undefined = {
            autoRefresh: true,
            interval: 6000,
            timeLeft: 70
        }
    ) {
        return options;
    }
}

export { uKeycloak };
