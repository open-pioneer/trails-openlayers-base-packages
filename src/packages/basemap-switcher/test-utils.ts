// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { MapConfig, MapConfigProvider, MapRegistry } from "@open-pioneer/map";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import Stamen from "ol/source/Stamen";
import { createService } from "@open-pioneer/test-utils/services";
import { MapRegistryImpl } from "@open-pioneer/map/MapRegistryImpl";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { screen, waitFor } from "@testing-library/react";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}
export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const mapConfig: MapConfig = {
        initialView: {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        },
        projection: "EPSG:3857",
        layers: [
            {
                id: "b-1",
                title: "OSM",
                isBaseLayer: true,
                visible: false,
                layer: new TileLayer({
                    source: new OSM()
                })
            },
            {
                id: "b-2",
                title: "Toner",
                isBaseLayer: true,
                visible: true,
                layer: new TileLayer({
                    source: new Stamen({ layer: "toner" })
                })
            }
        ]
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
