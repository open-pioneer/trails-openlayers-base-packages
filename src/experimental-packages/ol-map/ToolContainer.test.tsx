// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
/**
 * @vitest-environment jsdom
 */
import { OlMapConfigurationProvider } from "./api";
import { MapContainer } from "./MapContainer";
import { OlMapRegistry } from "./services";
import { Service, ServiceOptions } from "@open-pioneer/runtime";
import {
    PackageContextProvider,
    PackageContextProviderProps
} from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { render, screen, waitFor } from "@testing-library/react";
import { MapOptions } from "ol/Map";
import { expect, it } from "vitest";
import { ToolContainer } from "./ToolContainer";
import { Box } from "@open-pioneer/chakra-integration";

// used to avoid a "ResizeObserver is not defined" error
global.ResizeObserver = require("resize-observer-polyfill");

class MapConfigProvider implements OlMapConfigurationProvider {
    mapId = "default";
    mapOptions: MapOptions = {};

    constructor(options: ServiceOptions) {
        if (options.properties.mapOptions) {
            this.mapOptions = options.properties.mapOptions as MapOptions;
        }
        if (options.properties.mapId) {
            this.mapId = options.properties.mapId as string;
        }
    }

    getMapOptions(): Promise<MapOptions> {
        return Promise.resolve(this.mapOptions);
    }
}

async function createOlMapRegistry(mapId: string, mapOptions: MapOptions) {
    const mapConfigProvider = await createService(MapConfigProvider, {
        properties: {
            mapOptions: mapOptions,
            mapId
        }
    });
    return await createService(OlMapRegistry, {
        references: {
            providers: [mapConfigProvider]
        }
    });
}

function createPackageContextProviderProps(
    service: Service<OlMapRegistry>
): PackageContextProviderProps {
    return {
        services: {
            "ol-map.MapRegistry": service
        }
    };
}

it("should successfully create a tool container component", async () => {
    const mapId = "test";
    const mapOptions = {} as MapOptions;
    const service = await createOlMapRegistry(mapId, mapOptions);

    const { container } = render(
        <PackageContextProvider {...createPackageContextProviderProps(service)}>
            <div data-testid="base">
                <MapContainer mapId={mapId}>
                    <ToolContainer></ToolContainer>
                </MapContainer>
            </div>
        </PackageContextProvider>
    );

    // assert tool container is mounted
    await waitFor(async () => {
        const domElement = await screen.findByTestId("base");
        const toolContainer = domElement.querySelector(".tool-container");
        if (!toolContainer) {
            throw new Error("tool container not rendered");
        }
        return domElement;
    });

    // check tool container box is available
    const toolContainer = container.querySelector(".tool-container");
    if (!toolContainer) {
        throw new Error("tool container not rendered");
    } else {
        expect(toolContainer).toBeInstanceOf(HTMLDivElement);
        expect(toolContainer).toMatchSnapshot();
    }
});

it("should successfully create a tool container component with additional css classes", async () => {
    const mapId = "test";
    const mapOptions = {} as MapOptions;
    const service = await createOlMapRegistry(mapId, mapOptions);

    const { container } = render(
        <PackageContextProvider {...createPackageContextProviderProps(service)}>
            <div data-testid="base">
                <MapContainer mapId={mapId}>
                    <ToolContainer className="test test1 test2"></ToolContainer>
                </MapContainer>
            </div>
        </PackageContextProvider>
    );

    // assert tool container is mounted
    await waitFor(async () => {
        const domElement = await screen.findByTestId("base");
        const toolContainer = domElement.querySelector(".tool-container");
        if (!toolContainer) {
            throw new Error("tool container not rendered");
        }
        return domElement;
    });

    // check tool container box is available with additional css classes
    const toolContainer = container.querySelector(".tool-container");
    if (!toolContainer) {
        throw new Error("tool container not rendered");
    } else {
        expect(toolContainer).toBeInstanceOf(HTMLDivElement);
        expect(toolContainer.classList.contains("test")).toBe(true);
        expect(toolContainer.classList.contains("test1")).toBe(true);
        expect(toolContainer.classList.contains("test2")).toBe(true);
        expect(toolContainer.classList.contains("test3")).not.toBe(true);
    }
});

it('should successfully create a tool container component with prop `position="top-left"`', async () => {
    const mapId = "test";
    const mapOptions = {} as MapOptions;
    const service = await createOlMapRegistry(mapId, mapOptions);

    const { container } = render(
        <PackageContextProvider {...createPackageContextProviderProps(service)}>
            <div data-testid="base">
                <MapContainer mapId={mapId}>
                    <ToolContainer position="top-left"></ToolContainer>
                </MapContainer>
            </div>
        </PackageContextProvider>
    );

    // assert tool container is mounted
    await waitFor(async () => {
        const domElement = await screen.findByTestId("base");
        const toolContainer = domElement.querySelector(".tool-container");
        if (!toolContainer) {
            throw new Error("tool container not rendered");
        }
        return domElement;
    });

    // check tool container box is available with style for prop `position`
    const toolContainer = container.querySelector(".tool-container");
    if (!toolContainer) {
        throw new Error("tool container not rendered");
    } else {
        const styles = window.getComputedStyle(toolContainer);
        expect(styles.top).toBe("0px");
        expect(styles.left).toBe("0px");
    }
});

it('should successfully create a tool container component with prop `position="bottom-right"`', async () => {
    const mapId = "test";
    const mapOptions = {} as MapOptions;
    const service = await createOlMapRegistry(mapId, mapOptions);

    const { container } = render(
        <PackageContextProvider {...createPackageContextProviderProps(service)}>
            <div data-testid="base">
                <MapContainer mapId={mapId}>
                    <ToolContainer position="bottom-right"></ToolContainer>
                </MapContainer>
            </div>
        </PackageContextProvider>
    );

    // assert tool container is mounted
    await waitFor(async () => {
        const domElement = await screen.findByTestId("base");
        const toolContainer = domElement.querySelector(".tool-container");
        if (!toolContainer) {
            throw new Error("tool container not rendered");
        }
        return domElement;
    });

    // check tool container box is available with style for prop `position`
    const toolContainer = container.querySelector(".tool-container");
    if (!toolContainer) {
        throw new Error("tool container not rendered");
    } else {
        const styles = window.getComputedStyle(toolContainer);
        expect(styles.right).toBe("0px");
        // improvement import / export attributionGap
        expect(styles.bottom).toBe("30px"); // "30px" (attributionGap) if no verticalGap is configured
    }
});

it('should successfully create a tool container component with props `position="bottom-right"` and `horizontalGap={30} verticalGap={10}`', async () => {
    const mapId = "test";
    const mapOptions = {} as MapOptions;
    const service = await createOlMapRegistry(mapId, mapOptions);

    const { container } = render(
        <PackageContextProvider {...createPackageContextProviderProps(service)}>
            <div data-testid="base">
                <MapContainer mapId={mapId}>
                    <ToolContainer
                        position="bottom-right"
                        horizontalGap={30}
                        verticalGap={10}
                    ></ToolContainer>
                </MapContainer>
            </div>
        </PackageContextProvider>
    );

    // assert tool container is mounted
    await waitFor(async () => {
        const domElement = await screen.findByTestId("base");
        const toolContainer = domElement.querySelector(".tool-container");
        if (!toolContainer) {
            throw new Error("tool container not rendered");
        }
        return domElement;
    });

    // check tool container box is available with style for props `position`, `horizontalGap` and `verticalGap`
    const toolContainer = container.querySelector(".tool-container");
    if (!toolContainer) {
        throw new Error("tool container not rendered");
    } else {
        const styles = window.getComputedStyle(toolContainer);
        expect(styles.right).toBe("30px");
        expect(styles.bottom).toBe("10px");
    }
});

it("should successfully create a tool container component with ReactNode as children", async () => {
    const mapId = "test";
    const mapOptions = {} as MapOptions;
    const service = await createOlMapRegistry(mapId, mapOptions);

    const { container } = render(
        <PackageContextProvider {...createPackageContextProviderProps(service)}>
            <div data-testid="base">
                <MapContainer mapId={mapId}>
                    <ToolContainer>
                        <Box className="chakra-ui-box">Chakra UI Box</Box>
                    </ToolContainer>
                </MapContainer>
            </div>
        </PackageContextProvider>
    );

    // assert tool container is mounted
    await waitFor(async () => {
        const domElement = await screen.findByTestId("base");
        const toolContainer = domElement.querySelector(".tool-container");
        if (!toolContainer) {
            throw new Error("tool container not rendered");
        }
        return domElement;
    });

    // check tool container box is available with ReactNode as children
    const toolContainer = container.querySelector(".tool-container");
    if (!toolContainer) {
        throw new Error("tool container not rendered");
    } else {
        const div = await waitFor(async () => {
            const domElement = toolContainer.querySelector(".chakra-ui-box");
            if (!domElement) {
                throw new Error("child element in tool container not rendered");
            }
            return domElement;
        });

        expect(div.innerHTML).toBe("Chakra UI Box");
    }
});

// TODO
// zwei gleichzeitig einbinden
// maxH maxW -> import computePositionStyles
