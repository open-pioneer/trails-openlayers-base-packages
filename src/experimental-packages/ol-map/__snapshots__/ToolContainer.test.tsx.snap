// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should successfully create a tool container component 1`] = `
<div
  data-testid="base"
>
  <div
    class="map-container css-v2kfba"
    data-theme="light"
  >
    <div
      class="ol-viewport ol-touch"
      style="position: relative; overflow: hidden; width: 100%; height: 100%;"
    >
      <div
        class="ol-unselectable ol-layers"
        style="position: absolute; width: 100%; height: 100%; z-index: 0;"
      />
      <div
        class="ol-overlaycontainer"
        style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;"
      />
      <div
        class="ol-overlaycontainer-stopevent"
        style="position: absolute; z-index: 0; width: 100%; height: 100%; pointer-events: none;"
      >
        <div
          class="ol-zoom ol-unselectable ol-control"
          style="pointer-events: auto;"
        >
          <button
            class="ol-zoom-in"
            title="Zoom in"
            type="button"
          >
            +
          </button>
          <button
            class="ol-zoom-out"
            title="Zoom out"
            type="button"
          >
            –
          </button>
        </div>
        <div
          class="ol-rotate ol-unselectable ol-control ol-hidden"
          style="pointer-events: auto;"
        >
          <button
            class="ol-rotate-reset"
            title="Reset rotation"
            type="button"
          >
            <span
              class="ol-compass"
            >
              ⇧
            </span>
          </button>
        </div>
        <div
          class="ol-attribution ol-unselectable ol-control ol-collapsed"
          style="pointer-events: auto;"
        >
          <button
            aria-expanded="false"
            title="Attributions"
            type="button"
          >
            <span
              class="ol-attribution-expand"
            >
              i
            </span>
          </button>
          <ul />
        </div>
        <div
          class="tool-container css-lhbs93"
          data-theme="light"
        />
      </div>
    </div>
  </div>
</div>
`;
